Lib/statistics.py

```diff
@@ -305,28 +314,58 @@ def _fail_neg(values, errmsg='negative value'):
             raise StatisticsError(errmsg)
         yield x
 
-def _isqrt_frac_rto(n: int, m: int) -> float:
+
+def _integer_sqrt_of_frac_rto(n: int, m: int) -> int:
     """Square root of n/m, rounded to the nearest integer using round-to-odd."""
     # Reference: https://www.lri.fr/~melquion/doc/05-imacs17_1-expose.pdf
     a = math.isqrt(n // m)
     return a | (a*a*m != n)
 
-# For 53 bit precision floats, the _sqrt_frac() shift is 109.
+
+# For 53 bit precision floats, the _float_sqrt_of_frac() shift is 109.
 _sqrt_shift: int = 2 * sys.float_info.mant_dig + 3
 
-def _sqrt_frac(n: int, m: int) -> float:
+
+def _float_sqrt_of_frac(n: int, m: int) -> float:
     """Square root of n/m as a float, correctly rounded."""
     # See principle and proof sketch at: https://bugs.python.org/msg407078
     q = (n.bit_length() - m.bit_length() - _sqrt_shift) // 2
     if q >= 0:
-        numerator = _isqrt_frac_rto(n, m << 2 * q) << q
+        numerator = _integer_sqrt_of_frac_rto(n, m << 2 * q) << q
         denominator = 1
     else:
-        numerator = _isqrt_frac_rto(n << -2 * q, m)
+        numerator = _integer_sqrt_of_frac_rto(n << -2 * q, m)
         denominator = 1 << -q
     return numerator / denominator   # Convert to float
 
 
+def _decimal_sqrt_of_frac(n: int, m: int) -> Decimal:
+    """Square root of n/m as a float, correctly rounded."""
+    # Premise:  For decimal, computing (n/m).sqrt() can be off by 1 ulp.
+    # Method:   Check the result, moving up or down a step if needed.
+    if n <= 0:
```

Comment:
Should we be checking `m` here rather than `n`? It looks as though we multiply through by `m` below, so we need `m` to be positive for the inequalities to work.

Reply:
The case where *m* is zero is handled before the inequality test.  The `Decimal(n) / Decimal(m)` step raises *DivisionByZero* which is a subclass of *ZeroDivisionError*.  There is a test for this case.

Reply:
Yes, I was really more worried about the case m negative, not m zero. It seems more natural to have the normalization step `n, m = -n, -m` ensure that the *denominator* is positive than that the numerator: right now, we're effectively converting `-2 / 5` to `2 / (-5)`, and what we care about below for the maths to work is that `m` is positive, not that `n` is positive: you're implicitly converting the inequality `n / m > ((root + plus) / 2) ** 2` to the inequality `n > ((root + plus) / 2)**2 * m`, and that conversion is only valid if `m` is positive.

It doesn't matter, because if just one of `n` and `m` is negative then the `(Decimal(n) / Decimal(m)).sqrt()` step will fail (except in the case of really extreme `m` where the division rounds from something tiny and negative to negative zero); it just seems like a surprising choice and for me at least it made the code harder to read and reason about.